### ISSUE TITLE
[Feature] Python 3.10 wheel

### ISSUE THREAD

**Issue by austinnichols101** [ray-project/ray#19116]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Python 3.10 released: 

3.10.0 final: Monday, 2021-10-04

### Related issues

https://github.com/ray-project/ray/issues/11287

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by darcamo** [comment]
Any news on this?

Arch Linux just updated to Python 3.10 (3.10.1) and when trying to recreate my virtual environments ( using [poetry](https://python-poetry.org/)) it failed with error
```
RuntimeError

  Unable to find installation candidates for ray (1.9.0)
```
I suspect it is because of the wheels missing. Everything worked fine on Python 3.9.

**Comment by kyoungseoun-chung** [comment]
> Any news on this?
> 
> Arch Linux just updated to Python 3.10 (3.10.1) and when trying to recreate my virtual environments ( using [poetry](https://python-poetry.org/)) it failed with error
> 
> ```
> RuntimeError
> 
>   Unable to find installation candidates for ray (1.9.0)
> ```
> 
> I suspect it is because of the wheels missing. Everything worked fine on Python 3.9.

I also faced the same problem. And seems like definitely wheel is missing...

```python
# ray/python/setup.py:
SUPPORTED_PYTHONS = [(3, 6), (3, 7), (3, 8), (3, 9)]
```

Anyone working on this?

**Comment by mzaiprog** [comment]
Adding 3.10 support via patching:
`# ray/python/setup.py:`
`SUPPORTED_PYTHONS = [(3, 6), (3, 7), (3, 8), (3, 9)]`

to
`# ray/python/setup.py:`
`SUPPORTED_PYTHONS = [(3, 6), (3, 7), (3, 8), (3, 9), (3, 10)]`

Did the trick for me and let it build & install on Windows 11 by following the steps denoted here:
https://docs.ray.io/en/master/ray-contribute/development.html#building-ray-on-windows-full

Tensorflow inference in Actor instances called from ActorPool does work properly.

**Comment by dynamicwebpaige** [comment]
#### Note, for context: 

This issue is also a blocker for [FlyteKit](https://github.com/flyteorg/flytekit), a Python SDK for [Flyte](https://flyte.org/) -- which is used in production by Lyft, Spotify, and Freenome.

**Comment by dashsatish** [comment]
Any updates on the wheels for python3.10. Atleast Ray-Core is very much awaited for :)

**Comment by cmertin** [comment]
There does seem to be a patch for [arch](https://aur.archlinux.org/packages/python-ray) for python 3.10, wonder why they haven't released it for 3.10 officially yet

**Comment by siekmanj** [comment]
Any updates on this issue?

**Comment by lizelive** [comment]
3.10 :package: support would be very nice makes ray seem dead without it

**Comment by wumpus** [comment]
To explain the why of the previous comment: I use ray in the guts of some middleware that facilitates distributed computing for the black hole research group that published that image of the black hole in the middle of our galaxy yesterday.

I would like to support as many python versions as possible, especially the "current" one. But I don't know how to install ray for python 3.10, reliably across many versions of Linux. Which is what wheels do best.

So, what do I do? Tell people don't use 3.10, use 3.9 or 3.11?

**Comment by mzaiprog** [comment]
At least on Windows 11 I managed to build ray tag:1.12.0 using the attached patch

by following the instructions on: https://docs.ray.io/en/latest/ray-contribute/development.html

May be it helps on linux too.
 

Mit freundlichen Grüßen
Mike Ziebeck

 
 

Gesendet: Samstag, 14. Mai 2022 um 04:30 Uhr
Von: "Greg Lindahl" ***@***.***>
An: "ray-project/ray" ***@***.***>
Cc: "mzaiprog" ***@***.***>, "Comment" ***@***.***>
Betreff: Re: [ray-project/ray] [Feature] Python 3.10 wheel (#19116)


 

To explain the why of the previous comment: I use ray in the guts of some middleware that facilitates distributed computing for the black hole research group that published that image of the black hole in the middle of our galaxy yesterday.

I would like to support as many python versions as possible, especially the "current" one. But I don't know how to install ray for python 3.10, reliably across many versions of Linux. Which is what wheels do best.

So, what do I do? Tell people don't use 3.10, use 3.9 or 3.11?

**Comment by matthiaskoenig** [comment]
Hi all,
it would be great if this could be solved.

This is already one of the highest rated issues based on reactions (so the community wants this)
https://github.com/ray-project/ray/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc
and there are multiple open pull requests which seem to address this.

Any chance of solving this or explaining what the big issue is with python 3.10?
Best Matthias

**Comment by tomprimozic** [comment]
At the very least it could be mentioned on the website that Ray doesn't support Python 3.10.

Currently if I follow instructions on the website `pip install 'ray[data]'` I get:

```
ERROR: Could not find a version that satisfies the requirement ray (from versions: none)
ERROR: No matching distribution found for ray
```

**Comment by pcmoritz** [comment]
We are trying hard to make Python 3.10 wheels available for the Ray 1.13 release :)

**Comment by zndr27** [comment]
This would be much appreciated! I am also looking for python 3.10 support.

**Comment by pcmoritz** [comment]
I'm closing this now, 3.10 wheels are now available for Ray 1.13 which just got released :)

**Comment by matthiaskoenig** [comment]
Could anybody get this working? I have the feeling the 3.10 wheels have not been distributed to pypi. 
Also I can't find them in the daily builds either listed here https://docs.ray.io/en/master/ray-overview/installation.html#daily-releases-nightlies
I am getting `Could not find a version that satisfies the requirement ray (from versions: none)` trying a pip install
```bash
trip3:~/git/pkdb_data$ mkvirtualenv ray-py310 --python=python3.10
created virtual environment CPython3.10.4.final.0-64 in 228ms
  creator CPython3Posix(dest=/home/mkoenig/envs/ray-py310, clear=False, global=False)
  seeder FromAppData(download=False, pip=latest, setuptools=latest, wheel=latest, via=copy, app_data_dir=/home/mkoenig/.local/share/virtualenv/seed-app-data/v1.0.1)
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
virtualenvwrapper.user_scripts creating /home/mkoenig/envs/ray-py310/bin/predeactivate
virtualenvwrapper.user_scripts creating /home/mkoenig/envs/ray-py310/bin/postdeactivate
virtualenvwrapper.user_scripts creating /home/mkoenig/envs/ray-py310/bin/preactivate
virtualenvwrapper.user_scripts creating /home/mkoenig/envs/ray-py310/bin/postactivate
virtualenvwrapper.user_scripts creating /home/mkoenig/envs/ray-py310/bin/get_env_details

(ray-py310) trip3:~/git/pkdb_data$ pip install ray
ERROR: Could not find a version that satisfies the requirement ray (from versions: none)
ERROR: No matching distribution found for ray
```

**Comment by mzaiprog** [comment]
Same here on Windows 11:
 
Microsoft Windows [Version 10.0.22000.675]
(c) Microsoft Corporation. Alle Rechte vorbehalten.

c:\Temp\a\ray>python -m venv .
c:\Temp\a\ray>call .\Scripts\activate.bat

(ray) c:\Temp\a\ray>python -m pip install --upgrade pip
Requirement already satisfied: pip in c:\temp\a\ray\lib\site-packages (22.0.4)
Collecting pip
  Using cached pip-22.1.2-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.0.4
    Uninstalling pip-22.0.4:
      Successfully uninstalled pip-22.0.4
Successfully installed pip-22.1.2

(ray) c:\Temp\a\ray>pip install wheel
Collecting wheel
  Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel
Successfully installed wheel-0.37.1

(ray) c:\Temp\a\ray>pip install ray
ERROR: Could not find a version that satisfies the requirement ray (from versions: none)
ERROR: No matching distribution found for ray

(ray) c:\Temp\a\ray>python -V
Python 3.10.4

(ray) c:\Temp\a\ray>pip list
Package    Version
---------- -------
pip        22.1.2
setuptools 58.1.0
wheel      0.37.1

**Comment by StevenKessler** [comment]
I can install ray, but it does not work for me with python 3.10:

`TypeError: As of 3.10, the *loop* parameter was removed from Lock() since it is no longer necessary`

**Comment by acxz** [comment]
@pcmoritz this issue seems to be prematurely closed with only partial support for 3.10 can be open it up again until #21221 is merged in?

**Comment by acxz** [comment]
Completed!

**Comment by scv119** [comment]
thanks @acxz!

**Comment by fehrin1** [comment]
At least for Windows this is not resolved, with Python 3.10 on Windows 10 20H2 i get
pip install -U ray
ERROR: Could not find a version that satisfies the requirement ray (from versions: none)
ERROR: No matching distribution found for ray
Is there any progress (to be expected)?

**Comment by fehrin1** [comment]
What works is
pip install -U ray-cpp
Collecting ray-cpp
  Downloading ray_cpp-2.0.0-cp310-cp310-win_amd64.whl (18.5 MB)
But this is still not ray.

**Comment by fehrin1** [comment]
On https://pypi.org/project/ray/#files i can see there are cp310 packages for Linux and MacOS but not for Windows, are they to be supplied in the foreseeable future?

**Comment by AndreMPCosta** [comment]
Same problem:
```

pip install -U ray

ERROR: Could not find a version that satisfies the requirement ray (from versions: none)
ERROR: No matching distribution found for ray
```

Windows 11, Python 3.10

**Comment by rickyyx** [comment]
> On https://pypi.org/project/ray/#files i can see there are cp310 packages for Linux and MacOS but not for Windows, are they to be supplied in the foreseeable future?

Windows py310 wheels will be supported in the next release. 

For mitigation - nightly windows python310 wheel is available: https://docs.ray.io/en/latest/ray-overview/installation.html#daily-releases-nightlies

**Comment by rickyyx** [comment]
Fixed the missing 2.2.0 window wheel for py3.10 and uploaded to pypi. https://pypi.org/project/ray/2.2.0/#files
